File: tools/shapeverify.mjs
#!/usr/bin/env node
/**
 * Prove every bundled country outline is actually that country.
 *
 * Rwanda shipped showing Saudi Arabia. The mistake was upstream — mapsicon's
 * own rw folder contains the Saudi file, byte for byte — so downloading by
 * country code was never going to catch it: the file was exactly where it
 * belonged and simply held the wrong country. A duplicate check (see
 * shapecheck.mjs) catches that case, but not the general one, because an
 * outline can be wrong without happening to duplicate another one in the set.
 *
 * So each outline is compared against independent geometry. Natural Earth's
 * public-domain polygons and the bundled SVG are each rasterised to a small
 * grid and compared by overlap.
 *
 * Three things had to be got right, and each wrong answer was informative:
 *
 *   An absolute similarity threshold flagged 32 countries including Croatia,
 *   Sweden and Mexico. The two sources do not share a projection, so a correct
 *   country scores modestly and no threshold separates it from a wrong one.
 *
 *   Ranking against all references reported France as the 174th best match for
 *   France. Whole-country comparison is dominated by overseas territories,
 *   date-line splits and island scatter, so only the largest landmass is
 *   compared: no two countries share a mainland.
 *
 *   It then claimed Italy resembled 168 countries more than itself, which is
 *   absurd enough to be a bug rather than a finding — and was. Every mapsicon
 *   file wraps its paths in transform="translate(0,1024) scale(0.1,-0.1)", and
 *   ignoring it compared every country upside down.
 *
 * The verdict is the gap: how much better some other country fits than your
 * own. Rwanda scored 0.48 for itself and 0.89 for Saudi Arabia. Nothing correct
 * came near that.
 *
 * Written in Node rather than Python so the build depends on one runtime. The
 * Python version used int.bit_count(), which needs 3.10, and failed the deploy.
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const GRID = 56;   /* raster size: enough to tell countries apart, quick enough */
const FLAT = 6;    /* straight segments per cubic curve */
const GAP = 0.30;  /* how much better another country must fit to be a verdict */

const HERE = path.dirname(fileURLToPath(import.meta.url));
const DIR = 'assets/img/shapes';

/**
 * Countries the two sources genuinely disagree about, each checked by eye.
 * These are places where "which landmass is the country" is not a question the
 * two datasets answer the same way — not wrong outlines.
 */
const KNOWN = {
  aq: 'Antarctica: circumpolar, so the reference collapses to a line',
  gl: 'Greenland: near the pole, where the reference projection flattens it',
  my: 'Malaysia: peninsula and Borneo, and the sources pick different mainlands',
  ru: 'Russia: crosses the date line, which splits the reference in two',
  eh: 'Western Sahara: disputed, and the two datasets draw different borders'
};

/* ------------------------------------------------------------------ */
/* SVG paths                                                           */
/* ------------------------------------------------------------------ */

const TOKEN = /[A-Za-z]|-?\d*\.?\d+(?:[eE][-+]?\d+)?/g;

/** Flatten one path to closed rings. Only what mapsicon emits: M/m L/l c z. */
function parsePath(d) {
  const t = d.match(TOKEN) || [];
  const rings = [];
  let ring = [];
  let x = 0, y = 0, sx = 0, sy = 0, i = 0, cmd = null;
  const num = () => Number(t[i++]);
  const flush = () => { if (ring.length > 2) rings.push(ring); ring = []; };

  while (i < t.length) {
    if (/^[A-Za-z]$/.test(t[i])) {
      cmd = t[i++];
      if (cmd === 'z' || cmd === 'Z') { flush(); x = sx; y = sy; continue; }
    }
    if (cmd === null) { i += 1; continue; }
    if (cmd === 'M' || cmd === 'm') {
      flush();
      const dx = num(), dy = num();
      if (cmd === 'M') { x = dx; y = dy; } else { x += dx; y += dy; }
      sx = x; sy = y;
      ring.push([x, y]);
      cmd = cmd === 'M' ? 'L' : 'l';
    } else if (cmd === 'L' || cmd === 'l') {
      const dx = num(), dy = num();
      if (cmd === 'L') { x = dx; y = dy; } else { x += dx; y += dy; }
      ring.push([x, y]);
    } else if (cmd === 'C' || cmd === 'c') {
      const p = [num(), num(), num(), num(), num(), num()];
      const [x1, y1, x2, y2, x3, y3] = cmd === 'c'
        ? [x + p[0], y + p[1], x + p[2], y + p[3], x + p[4], y + p[5]]
        : p;
      for (let k = 1; k <= FLAT; k += 1) {
        const s = k / FLAT, u = 1 - s;
        ring.push([
          u * u * u * x + 3 * u * u * s * x1 + 3 * u * s * s * x2 + s * s * s * x3,
          u * u * u * y + 3 * u * u * s * y1 + 3 * u * s * s * y2 + s * s * s * y3
        ]);
      }
      x = x3; y = y3;
    } else if (cmd === 'H' || cmd === 'h') {
      const dx = num(); x = cmd === 'H' ? dx : x + dx; ring.push([x, y]);
    } else if (cmd === 'V' || cmd === 'v') {
      const dy = num(); y = cmd === 'V' ? dy : y + dy; ring.push([x, y]);
    } else { i += 1; }
  }
  flush();
  return rings;
}

/** Rings from a file, with the group transform applied — the minus sign matters. */
function svgRings(file) {
  const src = fs.readFileSync(file, 'utf8');
  let tx = 0, ty = 0, sx = 1, sy = 1;
  const tm = src.match(/transform="([^"]+)"/);
  if (tm) {
    const tr = tm[1].match(/translate\(\s*(-?[\d.]+)\s*,\s*(-?[\d.]+)\s*\)/);
    if (tr) { tx = Number(tr[1]); ty = Number(tr[2]); }
    const sc = tm[1].match(/scale\(\s*(-?[\d.]+)\s*(?:,\s*(-?[\d.]+)\s*)?\)/);
    if (sc) { sx = Number(sc[1]); sy = sc[2] === undefined ? sx : Number(sc[2]); }
  }
  const out = [];
  for (const m of src.matchAll(/\sd="([^"]+)"/g)) {
    for (const ring of parsePath(m[1])) {
      out.push(ring.map(([px, py]) => [tx + sx * px, ty + sy * py]));
    }
  }
  return out;
}

/* ------------------------------------------------------------------ */
/* Rasterising                                                         */
/* ------------------------------------------------------------------ */

const area = (ring) => {
  let a = 0;
  for (let i = 0; i < ring.length; i += 1) {
    const [x1, y1] = ring[i];
    const [x2, y2] = ring[(i + 1) % ring.length];
    a += x1 * y2 - x2 * y1;
  }
  return Math.abs(a) / 2;
};

/** Only the largest landmass. See the header for why. */
const mainland = (rings) => (rings.length
  ? [rings.reduce((best, r) => (area(r) > area(best) ? r : best))] : []);

const WORDS = Math.ceil((GRID * GRID) / 32);

/** Normalise to fill the grid, scanline fill, pack to bits. */
function raster(ringsIn) {
  const rings = mainland(ringsIn);
  const pts = rings.flat();
  if (pts.length < 3) return null;
  let minX = Infinity, maxX = -Infinity, minY = Infinity, maxY = -Infinity;
  for (const [px, py] of pts) {
    if (px < minX) minX = px; if (px > maxX) maxX = px;
    if (py < minY) minY = py; if (py > maxY) maxY = py;
  }
  const w = maxX - minX, h = maxY - minY;
  if (!(w > 0) || !(h > 0)) return null;
  const k = (GRID - 2) / Math.max(w, h);
  const ox = (GRID - w * k) / 2 - minX * k;
  const oy = (GRID - h * k) / 2 - minY * k;
  const scaled = rings.map((r) => r.map(([px, py]) => [px * k + ox, py * k + oy]));

  const out = new Uint32Array(WORDS);
  for (let row = 0; row < GRID; row += 1) {
    const yc = row + 0.5;
    const hits = [];
    for (const r of scaled) {
      for (let j = 0; j < r.length; j += 1) {
        const [x1, y1] = r[j];
        const [x2, y2] = r[(j + 1) % r.length];
        if ((y1 <= yc && yc < y2) || (y2 <= yc && yc < y1)) {
          hits.push(x1 + ((yc - y1) * (x2 - x1)) / (y2 - y1));
        }
      }
    }
    hits.sort((a, b) => a - b);
    for (let j = 0; j + 1 < hits.length; j += 2) {
      const a = Math.max(0, Math.ceil(hits[j] - 0.5));
      const b = Math.min(GRID - 1, Math.floor(hits[j + 1] - 0.5));
      for (let c = a; c <= b; c += 1) {
        const bit = row * GRID + c;
        out[bit >>> 5] |= 1 << (bit & 31);
      }
    }
  }
  return out;
}

const popcount = (v) => {
  v -= (v >>> 1) & 0x55555555;
  v = (v & 0x33333333) + ((v >>> 2) & 0x33333333);
  return (((v + (v >>> 4)) & 0x0f0f0f0f) * 0x01010101) >>> 24;
};

function iou(a, b) {
  let inter = 0, union = 0;
  for (let i = 0; i < WORDS; i += 1) {
    inter += popcount(a[i] & b[i]);
    union += popcount(a[i] | b[i]);
  }
  return union ? inter / union : 0;
}

/* ------------------------------------------------------------------ */

const geoRings = (entry) =>
  entry.rings.flatMap((poly) => poly.map((ring) => ring.map(([lon, lat]) => [lon, -lat])));

const refPath = path.join(HERE, 'reference', 'countries.json');
if (!fs.existsSync(refPath)) {
  console.log(`reference geometry not found: ${refPath}`);
  process.exit(1);
}
const ref = JSON.parse(fs.readFileSync(refPath, 'utf8'));
const data = JSON.parse(fs.readFileSync('data/fun/shapes.json', 'utf8'));
const names = Object.fromEntries(data.countries.map((c) => [c.code, c.name]));

const refs = [];
for (const entry of ref.countries) {
  const g = raster(geoRings(entry));
  if (g) refs.push([entry.iso, g]);
}
const haveRef = new Set(refs.map((r) => r[0]));

const misses = [];
let checked = 0, noRef = 0;
for (const c of data.countries) {
  const file = `${DIR}/${c.code}.svg`;
  if (!fs.existsSync(file)) continue;
  if (!haveRef.has(c.code)) { noRef += 1; continue; }
  const mine = raster(svgRings(file));
  if (!mine) { noRef += 1; continue; }
  checked += 1;
  let bestScore = -1, bestCode = '', own = 0;
  for (const [rc, r] of refs) {
    const s = iou(mine, r);
    if (rc === c.code) own = s;
    if (s > bestScore) { bestScore = s; bestCode = rc; }
  }
  if (bestCode !== c.code) {
    misses.push({ gap: bestScore - own, code: c.code, name: c.name,
      bestCode, bestName: names[bestCode] || bestCode, bestScore, own });
  }
}

misses.sort((a, b) => b.gap - a.gap);
const wrong = misses.filter((m) => m.gap >= GAP && !KNOWN[m.code]);
const allowed = misses.filter((m) => m.gap >= GAP && KNOWN[m.code]);

console.log(`${checked} outlines matched against ${refs.length} reference countries, `
  + `${noRef} had no reference`);

if (misses.length) {
  console.log('\nWidest gaps between how well an outline fits another country and its own:');
  for (const m of misses.slice(0, 8)) {
    const flag = m.gap >= GAP ? (KNOWN[m.code] ? 'known' : 'WRONG') : '     ';
    console.log(`  ${flag} +${m.gap.toFixed(2)}  ${m.code} (${m.name}) fits itself `
      + `${m.own.toFixed(2)}, but ${m.bestCode} (${m.bestName}) ${m.bestScore.toFixed(2)}`);
  }
}
if (allowed.length) {
  console.log('\nAllowed through, each checked by eye:');
  allowed.forEach((m) => console.log(`  ok  ${m.code} — ${KNOWN[m.code]}`));
}
if (wrong.length) {
  console.log(`\nERRORS (${wrong.length}):`);
  wrong.forEach((m) => console.log(`  x ${m.code} (${m.name}) is drawn as `
    + `${m.bestCode} (${m.bestName})`));
  process.exit(1);
}
console.log(`\nNo outline fits another country more than ${GAP} better than its own.`);

File: tools/shapeverify.py
#!/usr/bin/env python3
"""Prove every bundled country outline is actually that country.

Rwanda shipped showing Saudi Arabia. The mistake was upstream -- mapsicon's own
rw folder contains the Saudi outline -- which means trusting the file name was
never enough, and a duplicate check alone would not be either: a country can be
given the wrong shape without that shape happening to duplicate another one in
the set.

So each outline is checked against independent geometry. Natural Earth's
public-domain country polygons are rasterised to a small grid, the bundled SVG
is rasterised the same way, and the two are compared by intersection over union.
A country drawn as itself scores high whatever projection either source used; a
country drawn as somewhere else does not.

The reference lives in tools/reference/countries.json: Natural Earth 1:110m,
stripped to ISO code and geometry and rounded, so the check runs on every build
without carrying an 800KB GeoJSON around.

Usage:  python3 tools/shapeverify.py
"""

import json, re, sys, os, glob, math

GRID = 56          # raster size; big enough to tell countries apart, small enough to be quick
FLAT = 6           # segments per cubic curve

# ---------------------------------------------------------------- svg paths

NUM = re.compile(r'-?\d*\.?\d+(?:[eE][-+]?\d+)?')

def parse_path(d):
    """Flatten an SVG path to a list of closed rings of (x, y).

    Only what mapsicon actually emits: M/m, L/l, c, z. Cubics are flattened to
    straight segments, which is plenty at this raster size.
    """
    toks = re.findall(r'[A-Za-z]|' + NUM.pattern, d)
    rings, ring = [], []
    x = y = 0.0
    sx = sy = 0.0
    i, cmd = 0, None
    def num():
        nonlocal i
        v = float(toks[i]); i += 1; return v
    while i < len(toks):
        t = toks[i]
        if re.fullmatch(r'[A-Za-z]', t):
            cmd = t; i += 1
            if cmd in 'zZ':
                if len(ring) > 2: rings.append(ring)
                ring = []; x, y = sx, sy
                continue
        if cmd is None:
            i += 1; continue
        if cmd in 'Mm':
            if len(ring) > 2: rings.append(ring)
            ring = []
            dx, dy = num(), num()
            x, y = (dx, dy) if cmd == 'M' else (x + dx, y + dy)
            sx, sy = x, y
            ring.append((x, y))
            cmd = 'L' if cmd == 'M' else 'l'
        elif cmd in 'Ll':
            dx, dy = num(), num()
            x, y = (dx, dy) if cmd == 'L' else (x + dx, y + dy)
            ring.append((x, y))
        elif cmd in 'Cc':
            p = [num() for _ in range(6)]
            if cmd == 'c':
                x1, y1, x2, y2, x3, y3 = (x+p[0], y+p[1], x+p[2], y+p[3], x+p[4], y+p[5])
            else:
                x1, y1, x2, y2, x3, y3 = p
            for k in range(1, FLAT + 1):
                t0 = k / FLAT; u = 1 - t0
                bx = u*u*u*x + 3*u*u*t0*x1 + 3*u*t0*t0*x2 + t0**3*x3
                by = u*u*u*y + 3*u*u*t0*y1 + 3*u*t0*t0*y2 + t0**3*y3
                ring.append((bx, by))
            x, y = x3, y3
        elif cmd in 'Hh':
            dx = num(); x = dx if cmd == 'H' else x + dx; ring.append((x, y))
        elif cmd in 'Vv':
            dy = num(); y = dy if cmd == 'V' else y + dy; ring.append((x, y))
        else:
            i += 1
    if len(ring) > 2: rings.append(ring)
    return rings

def svg_rings(path):
    """Rings from the file, with the group transform applied.

    Every mapsicon file wraps its paths in
    transform="translate(0,1024) scale(0.1,-0.1)". Ignoring it mirrors the
    country vertically, which is why the first run of this tool reported that
    Italy looked less like Italy than 168 other countries did. Normalisation
    absorbs the translate and the scale; the minus sign is the part that
    matters.
    """
    src = open(path, encoding='utf-8').read()
    tx = ty = 0.0
    sx = sy = 1.0
    m = re.search(r'transform="([^"]+)"', src)
    if m:
        t = re.search(r'translate\(\s*(-?[\d.]+)\s*,\s*(-?[\d.]+)\s*\)', m.group(1))
        if t:
            tx, ty = float(t.group(1)), float(t.group(2))
        c = re.search(r'scale\(\s*(-?[\d.]+)\s*(?:,\s*(-?[\d.]+)\s*)?\)', m.group(1))
        if c:
            sx = float(c.group(1))
            sy = float(c.group(2)) if c.group(2) is not None else sx
    rings = []
    for d in re.findall(r'\sd="([^"]+)"', src):
        for ring in parse_path(d):
            rings.append([(tx + sx * px, ty + sy * py) for px, py in ring])
    return rings

# ---------------------------------------------------------------- rasterise

def area(ring):
    a = 0.0
    for i in range(len(ring)):
        x1, y1 = ring[i]; x2, y2 = ring[(i + 1) % len(ring)]
        a += x1 * y2 - x2 * y1
    return abs(a) / 2

def biggest(rings):
    """Only the largest landmass.

    Comparing whole countries did not work. France carries its overseas
    departments, Russia is split across the date line, and Japan, the
    Philippines and New Zealand are scattered islands -- so the bounding box
    that everything is normalised into is set by the far-flung pieces rather
    than the country, and the two sources disagree about which pieces to
    include. France ranked 174th for itself. Comparing mainland to mainland
    removes the disagreement without weakening the test: no two countries have
    the same mainland.
    """
    return [max(rings, key=area)] if rings else []

def raster(rings, n=GRID):
    """Normalise to fill an n x n box, keeping aspect, then scanline fill."""
    rings = biggest(rings)
    pts = [p for r in rings for p in r]
    if len(pts) < 3:
        return None
    xs = [p[0] for p in pts]; ys = [p[1] for p in pts]
    w = max(xs) - min(xs); h = max(ys) - min(ys)
    if w <= 0 or h <= 0:
        return None
    k = (n - 2) / max(w, h)
    ox = (n - w * k) / 2 - min(xs) * k
    oy = (n - h * k) / 2 - min(ys) * k
    scaled = [[((px * k) + ox, (py * k) + oy) for px, py in r] for r in rings]

    grid = bytearray(n * n)
    for row in range(n):
        yc = row + 0.5
        xs_hit = []
        for r in scaled:
            m = len(r)
            for j in range(m):
                x1, y1 = r[j]; x2, y2 = r[(j + 1) % m]
                if (y1 <= yc < y2) or (y2 <= yc < y1):
                    xs_hit.append(x1 + (yc - y1) * (x2 - x1) / (y2 - y1))
        xs_hit.sort()
        for j in range(0, len(xs_hit) - 1, 2):
            a = max(0, int(math.ceil(xs_hit[j] - 0.5)))
            b = min(n - 1, int(math.floor(xs_hit[j + 1] - 0.5)))
            for c in range(a, b + 1):
                grid[row * n + c] = 1
    return grid

def bits(grid):
    """Pack a raster into one integer, so overlap is two machine ops."""
    v = 0
    for i, on in enumerate(grid):
        if on:
            v |= 1 << i
    return v

def iou(a, b):
    u = (a | b).bit_count()
    return (a & b).bit_count() / u if u else 0.0

# ---------------------------------------------------------------- reference

def geo_rings(entry):
    """Lon/lat rings, y flipped so north is up on a screen."""
    return [[(p[0], -p[1]) for p in ring] for poly in entry['rings'] for ring in poly]

"""Countries the two sources genuinely disagree about, checked by eye and
allowed through. Each is a place where "which landmass is the country" is not a
question the two datasets answer the same way, not a wrong outline. Every one of
these was rendered from both sources and compared before being listed here."""
KNOWN = {
    'aq': 'Antarctica: circumpolar, so the reference collapses to a line',
    'my': 'Malaysia: peninsula and Borneo, and the sources pick different mainlands',
    'gl': 'Greenland: near the pole, where the reference projection flattens it',
    'eh': 'Western Sahara: disputed, and the two datasets draw different borders',
    'ru': 'Russia: crosses the date line, which splits the reference in two',
}

def main():
    ref_path = os.path.join(os.path.dirname(__file__), 'reference', 'countries.json')
    if not os.path.exists(ref_path):
        sys.exit(f'reference geometry not found: {ref_path}')
    ref = json.load(open(ref_path, encoding='utf-8'))
    by_iso = {c['iso']: c for c in ref['countries']}

    data = json.load(open('data/fun/shapes.json', encoding='utf-8'))
    names = {c['code']: c['name'] for c in data['countries']}

    """An absolute similarity threshold is the wrong test: mapsicon and Natural
    Earth do not share a projection, so a correct country can score modestly and
    a threshold either misses real errors or drowns in false ones. What does
    hold across projections is RANK. A country drawn as itself resembles itself
    more than it resembles any of the other 176. So each outline is scored
    against every reference and the question is simply whether its own country
    comes first."""
    refs = []
    for code, feat in by_iso.items():
        g = raster(geo_rings(feat))
        if g:
            refs.append((code, bits(g)))

    misses, unchecked, checked = [], [], 0
    for c in data['countries']:
        code = c['code']
        svg = f'assets/img/shapes/{code}.svg'
        if not os.path.exists(svg) or code not in by_iso:
            if os.path.exists(svg):
                unchecked.append(code)
            continue
        g = raster(svg_rings(svg))
        if not g:
            unchecked.append(code)
            continue
        mine = bits(g)
        scored = sorted(((iou(mine, r), rc) for rc, r in refs), reverse=True)
        checked += 1
        best = scored[0]
        rank = next(i for i, (_, rc) in enumerate(scored) if rc == code) + 1
        own = next(sc for sc, rc in scored if rc == code)
        """Rank alone is too blunt. Small compact countries genuinely resemble
        one another, and Norway, Russia, Canada, Greenland and Antarctica are
        fragmented or polar enough that the two sources disagree about their
        mainland no matter what. What separates a wrong shape from a merely
        ambiguous one is the GAP: how much better some other country fits than
        your own does. Rwanda scored 0.48 for itself and 0.89 for Saudi Arabia,
        a gap of 0.41, while nothing correct came close to that."""
        if rank != 1:
            misses.append((round(best[0] - own, 3), rank, code, names.get(code, '?'),
                           best[1], names.get(best[1], best[1]), best[0], own))

    print(f'{checked} outlines matched against {len(refs)} reference countries')
    print(f'{len(unchecked)} had no reference to compare with\n')

    GAP = 0.30
    wrong = [m for m in misses if m[0] >= GAP and m[2] not in KNOWN]
    allowed = [m for m in misses if m[0] >= GAP and m[2] in KNOWN]
    misses.sort(reverse=True)

    if misses:
        print('Widest gaps between how well an outline fits another country and its own:')
        for gap, rank, code, name, bcode, bname, bscore, own in misses[:10]:
            flag = 'WRONG' if gap >= GAP else '     '
            print(f'  {flag} +{gap:.2f}  {code} ({name}) fits itself {own:.2f}, '
                  f'but {bcode} ({bname}) {bscore:.2f}')

    if allowed:
        print('\nAllowed through, checked by eye:')
        for m in allowed:
            print(f'  ok  {m[2]} — {KNOWN[m[2]]}')

    if wrong:
        print(f'\n{len(wrong)} outline(s) are drawn as the wrong country:')
        for gap, rank, code, name, bcode, bname, bscore, own in wrong:
            print(f'  x {code} ({name}) is drawn as {bcode} ({bname})')
    else:
        print(f'\nNo outline fits another country more than {GAP} better than its own. '
              'None is drawn as the wrong country.')
    return wrong

if __name__ == '__main__':
    bad = main()
    sys.exit(1 if bad else 0)
